File: src/main/java/com/architecture_map/belarus/service/impl/ConstructionImageServiceImpl.java
package com.architecture_map.belarus.service.impl;

import com.architecture_map.belarus.dto.ConstructionImageDto;
import com.architecture_map.belarus.entity.image.ConstructionImage;
import com.architecture_map.belarus.mapper.ConstructionImageMapper;
import com.architecture_map.belarus.repository.ConstructionImageRepository;
import com.architecture_map.belarus.repository.ConstructionRepository;
import com.architecture_map.belarus.repository.ImageRepository;
import com.architecture_map.belarus.service.ConstructionImageService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.Set;

@Service
@RequiredArgsConstructor
public class ConstructionImageServiceImpl implements ConstructionImageService {

    private final ConstructionImageRepository constructionImageRepository;
    private final ConstructionImageMapper constructionImageMapper;
    private final ConstructionRepository constructionRepository;
    private final ImageRepository imageRepository;

    @Override
    public ConstructionImage create(ConstructionImageDto image) {
        return constructionImageRepository.save(constructionImageMapper.toConstructionImage(image));
    }

    @Override
    public Set<ConstructionImage> getRandomAndUniqueImages(String gotImagesId) {
        return constructionImageRepository.getRandomAndUniqueImages(parseStringToArrayOfInt(gotImagesId));
    }

    @Override
    public Set<ConstructionImage> getByConstructionArchitecturalStyleId(Integer architecturalStyleId) {
        return constructionImageRepository.getByConstructionArchitecturalStyleId(architecturalStyleId);
    }

    @Override
    public Set<ConstructionImage> getImagesWithSameArchitecturalStyleByConstructionIdAcrossImagesOfCurrentConstruction(Integer constructionId) {
        return constructionImageRepository.getImagesWithSameArchitecturalStyleByConstructionIdAcrossImagesOfCurrentConstruction(
                constructionId,
                constructionRepository.findById(constructionId).orElseThrow().getArchitecturalStyle().getId()
        );
    }

    @Override
    public boolean deleteById(Integer id) {

        if (constructionImageRepository.existsById(id)) {
            constructionImageRepository.deleteById(id);
            return true;
        }

        return false;
    }

    private int[] parseStringToArrayOfInt(String string) {

        int[] arrayOfInt = new int[]{0};

        if (string != null) {

            String[] arrayOfStrings = string.split(",");
            arrayOfInt = new int[arrayOfStrings.length];

            for (int i = 0; i < arrayOfStrings.length; i++) {
                arrayOfInt[i] = Integer.parseInt(arrayOfStrings[i]);
            }
        }

        return arrayOfInt;
    }
}

File: src/main/kotlin/by/architecturemap/belarus/service/impl/ConstructionImageServiceImpl.kt
package by.architecturemap.belarus.service.impl

import by.architecturemap.belarus.entity.ConstructionImage
import by.architecturemap.belarus.repository.jpa.ConstructionImageRepository
import by.architecturemap.belarus.service.ConstructionImageService
import by.architecturemap.belarus.service.ConstructionService
import org.springframework.stereotype.Service

@Service
class ConstructionImageServiceImpl(
    private val constructionImageRepository: ConstructionImageRepository,
    private val constructionService: ConstructionService
) : ConstructionImageService {

    override fun create(image: ConstructionImage): ConstructionImage = constructionImageRepository.save(image)

    override fun getRandomAndUniqueImages(usedImagesId: String?) =
        constructionImageRepository.getRandomAndUniqueImages(parseStringToIntArray(usedImagesId))

    override fun find(architecturalStyleId: Int): List<ConstructionImage> =
        constructionImageRepository.findByConstructionArchitecturalStyleId(architecturalStyleId)

    override fun getImagesOfConstructionsWithSameArchitecturalStyle(constructionId: Int): List<ConstructionImage> =
        constructionImageRepository.getImagesOfConstructionsWithSameArchitecturalStyleByConstructionId(
            constructionId,
            constructionService.find(constructionId).architecturalStyle.id
        )

    override fun delete(id: Int) {
        constructionImageRepository.deleteById(id)
    }

    private fun parseStringToIntArray(numbers: String?): IntArray =
        numbers
            ?.split(",")
            ?.map { it.toInt() }
            ?.toIntArray() ?: intArrayOf(0)
}
